torch/utils/hooks.py

```diff
@@ -227,7 +227,6 @@ def hook(_, grad_output):
                         if res is not None and not (isinstance(res, tuple) and all(el is None for el in res)):
                             raise RuntimeError("Backward hook for Modules where no input requires "
                                                "gradient should always return None or None for all gradients.")
-                    self.grad_outputs = None
```

Comment:
From https://github.com/pytorch/pytorch/pull/82788 it seems like if no inputs `require_grad` we need to set `self.grad_outputs = None` to avoid a ref cycle if `create_graph=True`. If any input requires grad this is handled by the hook registered on the input's `grad_fn` on L149.

n00b qn about the ref-cycle mentioned here:

https://github.com/pytorch/pytorch/blob/a17069684c65c2654726429f924b1ad8791ba35f/test/test_autograd.py#L7912-L7916

How can I access the `graph` mentioned?

The way I understand it, the hooks which might reference BackwardHook are registered to `input.grad_fn` and `output.grad_fn`, so why would `input.grad.grad_fn`/`output.grad.grad_fn` have references to the hook?

Reply:
Oups my bad!
Updated the PR to not re-introduce this cycle!

"graph" in this case is the whole autograd graph, you can traverse it by repeatedly calling `.next_functions` on the `autograd.graph.Node` returned by `grad_fn`.
This problem happens in the case of create_graph=True  where the gradient itself is requiring grad and building the graph.